### DISCUSSION TITLE
How to use the "go to definition" function, when meet the code like "addon.tab.traps.getBlockly()"?

### DISCUSSION THREAD

**Discussion by OriLiMu** [ScratchAddons/ScratchAddons#7540] [Help] [answered]
In the `addons/` folder when I meet the code `addon.tab.traps.getBlockly()` I want to check the definition of the `getBlockly`. It wont work. 

The `addon` in the code is a parameter passed into the function no declaration of the type of it. It may be the reason why vscode cant find the definition of it. But there is a big problem, if vscode don't know the type of the addon, vscode cant make the autocompletion of `addon` which is refered a lot times in the code  that is unacceptable, very likely to make a typo in the code. So there should be a way to define the type addon. But I don't know how yet .

So I want to know how to make "go to definition" work. 
And how to make the vscode autocompletion work for addon.

**Comment by mxmou** [comment, marked as answer]
Try this: https://github.com/ScratchAddons/ScratchAddons/pull/1339#issuecomment-761619407
The comment should be above the line that starts with `export default async function`.

**Reply by OriLiMu** [reply]
Hi in the file `Trap.js`

I want to set the type of tab in the constructor.
Follow the step of copilot I make the change below. But not work, do you know how to fix it? 

```js
import Listenable from "../common/Listenable.js";
/** I added
 * @typedef {import("./Tab.js").Tab} Tab
 */

/**
 * Manages object trapping.
 * @extends Listenable
 */
export default class Trap extends Listenable {
   /** I added
   * @param {Tab} tab
   */
  constructor(tab) {
    super();
    this._react_internal_key = undefined;
    this._isWWW = () => tab.clientVersion === "scratch-www";
    this._getEditorMode = () => this._isWWW() && tab.editorMode;
    this._waitForElement = tab.waitForElement.bind(tab);
    this._cache = Object.create(null);
  }
```

**Reply by mxmou** [reply]
The import on line 3 should be `import("./Tab.js").default`, not `import("./Tab.js").Tab`.

**Comment by WorldLanguages** [comment]
Note that the blockly object is not created by us, it's an object that is used internally by the Scratch editor, we simply provide that method to expose it. So at the moment we don't provide any documentation or type annotations for the properties inside the blockly object.
